File: editor_bin/package_ui.py
# -*- coding:utf-8 -*- 
README = \
"usepage:" \
"pyhton package_ui.py <module>  [target]" \
"module:" \
"   -load:  只是生成ui2.ori.lua" \
"   -pack:  生成png贴图以及lua文件" \
"target:" \
"   common: 默认打包目标"\
"   360: 打包360"\
"   efun: 打包efun"


import sys
import os
import re
import json

_abspath = os.path.abspath
_pjoin = os.path.join

modle = sys.argv[1]
target = "common"
if len(sys.argv)>=3:
    target = sys.argv[2]


def _run_cmd(cmd):
    print "_run_cmd -------------::", cmd
    os.system(cmd.encode('cp936'))

def _mkdir(name):
    path = _pjoin(pwd, name)
    if not os.path.isdir(path):
        _run_cmd('MKDIR "%s"' % path)
    return path


pwd = os.getcwd()

PYTHON = _pjoin(pwd, '..', 'tools', 'Python27', 'python.exe')

tex_json = {}


pack_dir = _abspath(_pjoin(pwd, "../", "sg_ui"))
base_res = _abspath(_pjoin(pwd, '..', 'sg_ui/common'))
sub_res_map = {
    '360' :  _abspath(_pjoin(pwd, '..', 'sg_ui/360')),
    'efun' : _abspath(_pjoin(pwd, '..', 'sg_ui/efun'))
}

tmp_pack = _abspath(_pjoin(pwd, "tmp_pack"))
res = tmp_pack

_mkdir('ui_output')
png = _pjoin(pwd, 'ui_output', "ui2%d.png")
tex = _pjoin(pwd, 'ui_output', "ui2%d.json")
tex2 = _pjoin(pwd, 'ui_output', "ui2")
out = _pjoin(pwd, 'ui_output', "ui2.ori.lua")

ui2_conver = _pjoin(pwd, "ui_output", "ui2.conver.lua")
ui2_manual = _pjoin(pwd, "ui_output", "ui2.manual.lua")
ui2 = _pjoin(pwd, "ui_output", "ui2.lua")
ep = _pjoin(pwd, "ui_output", "ui2")
epbin = _pjoin(pwd, "../tools/lua/epbin.lua")

def _rm_tmp_pack_dir():
    if os.path.exists(tmp_pack):
        _run_cmd('RD /Q/S "%s"' % tmp_pack)

def _have_file(dir):
    for file in os.listdir(dir):
        return True
    return False


def _gen_pack_dir():
    global res
    if sub_res_map.has_key(target):
        target_path = sub_res_map[target]
        if _have_file(target_path+"\\png"):
            _mkdir(res)
            _run_cmd('COPY /Y "%s\\png\\*.png" "%s\\"' % (base_res, res))
            _run_cmd('COPY /Y "%s\\png\\*.png" "%s\\"' % (target_path, res))
            return
    res = base_res + "\\png"

def _gen_jsons_dir():
    global res
    if sub_res_map.has_key(target):
        target_path = sub_res_map[target]
        if _have_file(target_path+"\\json"):
            _mkdir(res)
            _run_cmd('COPY /Y "%s\\json\\*.json" "%s\\"' % (base_res, res))
            _run_cmd('COPY /Y "%s\\json\\*.json" "%s\\"' % (target_path, res))
            return
    res = base_res + "\\json"

def _del_file(name):
    if os.path.isfile(name):
        _run_cmd('DEL /Q/F "%s"' % name)

# texture packer
def call_tex_pack(png, tex, res, redundant_files):
    TMP_FILE = _pjoin(pwd, 'ui_output', 'texturepacker.output.log')
    if os.path.isfile(TMP_FILE):
        _run_cmd('del /Q/F "%s"' % TMP_FILE)

    cmd = ' '.join([
        'TexturePacker',
        '--algorithm MaxRects',
        '--maxrects-heuristics Best',
        '--pack-mode Best',
        '--premultiply-alpha',
        '--sheet %s' % png,
        '--texture-format png',
        '--extrude 1',
        '--data %s' % tex,
        '--format json-array',
        '"%s"' % res,
        '2> %s' % TMP_FILE,
        ])
    _run_cmd(cmd)

    f = open(TMP_FILE, 'rb')
    lines = f.readlines()
    print "<<---------------------------------------------->>"
    for line in lines:
        print line.strip()
    print "<<<EOF>>>"
    f.close()

    _run_cmd('del /Q/F "%s"' % TMP_FILE)

    if lines and len(lines) >= 1:
        assert lines[0].strip() == 'TexturePacker:: error: Not all sprites could be packed into the texture!', lines[0].strip()
        for line in lines[1:]:
            m = re.match(r'^TexturePacker:: warning: (.*)$', line.strip())
            assert m, line.strip()
            redundant_files.append(m.group(1))
        print "FALSE!"
        return False

    assert not lines
    print "TRUE!"
    return True

def _tmpdir(name):
    path = _pjoin(pwd, 'ui_output', name)
    if os.path.isdir(path):
        _run_cmd('RD /Q/S "%s"' % path)
    _run_cmd('MKDIR "%s"' % path)
    print "_tmpdir ==============>", path
    return path

def pack_textures():
    _gen_pack_dir()
    png1 = png % 1
    tex1 = tex % 1
    redundant_files = []
    tex_json[0] = tex1
    if call_tex_pack(png1, tex1, res, redundant_files):
        _del_file('ui22.png')
        _del_file('ui22.json')
        return

    print "redundant_files:", redundant_files
    tmpdir1 = _tmpdir('tmp1')
    tmpdir2 = _tmpdir('tmp2')

    _run_cmd('COPY /Y "%s\\*.png" "%s\\"' % (res, tmpdir1))
    for name in redundant_files:
        _run_cmd('MOVE "%s\\%s" "%s\\"' % (tmpdir1, name, tmpdir2))

    png1 = png % 1
    tex1 = tex % 1
    redundant_files = []
    assert call_tex_pack(png1, tex1, tmpdir1, redundant_files)
    assert not redundant_files

    png2 = png % 2
    tex2 = tex % 2
    tex_json[1] = tex2
    redundant_files = []
    assert call_tex_pack(png2, tex2, tmpdir2, redundant_files)
    assert not redundant_files

    _run_cmd('RD /Q/S "%s"' % tmpdir1)
    _run_cmd('RD /Q/S "%s"' % tmpdir2)
    _rm_tmp_pack_dir()


def _json2map(json_file):
    handle = open(json_file)
    str = handle.read()
    handle.close()
    return json.loads(str)

def _get_pentagon_src(json_file):
    data = _json2map(json_file)
    for f in data["frames"]:
        if f["filename"] == "6450-gaoguang-1.png":
            frame = f["frame"]
            return [frame["x"]+2, frame["y"]+frame["h"]-2,  frame["x"]+frame["w"]-2, frame["y"]+frame["h"]-2,  frame["x"]+frame["w"]-2, frame["y"]+2,  frame["x"]+2, frame["y"]+2]
    return False


######################################### pentagon ########################################################
def get_pentagon():
    for k in tex_json:
        v = tex_json[k]
        src = _get_pentagon_src(v)
        if src:
            src_pos = ('"%d,%d, %d,%d, %d,%d, %d,%d"' % (src[0],src[1],  src[2],src[3], src[4],src[5], src[6],src[7]))
            tex = k
            return src_pos, tex
    assert(False)

def run_pentagon():
    src_pos, tex = get_pentagon()
    # gen char
    cmd = '%s ../pentagon_import/gen_pentagon.py %s %d ../pentagon_import/character.csv ../pentagon_import/pt_char.lua 19 %d'  % (PYTHON, src_pos, tex, 10527)
    _run_cmd(cmd)
    # gen hero
    cmd = '%s ../pentagon_import/gen_pentagon.py %s %d ../pentagon_import/hero.csv ../pentagon_import/pt_hero.lua 7 %d' % (PYTHON, src_pos, tex, 11527)
    _run_cmd(cmd)


######################################### gen ep ########################################################
def gen_ep():
    _del_file(ui2)
    _run_cmd("type %s >> %s" % (ui2_conver, ui2))
    _run_cmd("type %s >> %s" % (out, ui2))
    _run_cmd("type %s >> %s" % (ui2_manual, ui2))

    print("---------gen ios ep file.----------------")
    _run_cmd(r"..\tools\lua\lua %s -pd -png8 %s %s.ios" % (epbin, ui2, ep))
    print("---------gen android ep file.----------------")
    _run_cmd(r"..\tools\lua\lua %s -pd -pkmc %s %s.android" % (epbin, ui2, ep))


######################################### package ########################################################
def pack_load():
    _gen_jsons_dir()
    cmd = "cocpackage_load.exe %s %s %s" % (res, tex2, out)
    _run_cmd(cmd) 
    _rm_tmp_pack_dir()

######################################### merge ######################################################## 
def pack_merge():
    cmd = "type ..\\pentagon_import\\pt_char.lua >> %s" % out
    _run_cmd(cmd)
    cmd = " type ..\\pentagon_import\\pt_hero.lua >> %s" % out
    _run_cmd(cmd)

######################################### deff ########################################################
def  _gen_exist_filelist():
    deff = {}
    for k in tex_json:
        v = tex_json[k]
        data = _json2map(v)
        for frame_data in data['frames']:
            filename = frame_data['filename']
            fe = os.path.splitext(filename)
            deff[fe[0]] = True
    return deff

def _gen_update_filelist():
    up_file = []
    deff = _gen_exist_filelist()
    files = os.listdir(res)
    for f in files:
        fe = os.path.splitext(f)
        if fe[1] == '.png' and deff.has_key(fe[0]) != True:
            print('catch file:', f, deff.has_key(f))
            up_file.append(f)
    return up_file

def pack_up():
    up_file = _gen_update_filelist()
    tmpup = _tmpdir('tmp_up')
    for k in up_file:
        _run_cmd('COPY /Y "%s\\%s" "%s\\"' % (res, k, tmpup))
    
    redundant_files = []
    png_up = png % 3
    tex_up = tex % 3
    call_tex_pack(png_up, tex_up, tmpup, redundant_files)
    _run_cmd('RD /Q/S "%s"' % tmpup)

########################################################
######################## run ###########################
########################################################
if modle == '-load':
    tex_json[0] = tex % 1
    tex_json[1] = tex % 2
    run_pentagon()
    pack_load()
    pack_merge()
    gen_ep()

elif modle == '-pack':
    pack_textures()
    run_pentagon()
    pack_load()
    pack_merge()
    gen_ep()

elif modle == '-update':
    tex_json[0] = tex % 1
    tex_json[1] = tex % 2
    pack_up()
    run_pentagon()
    pack_load()
    pack_merge()
    gen_ep()

else  :
    print(README)

File: editor_bin/package_ui.lua
local README = [[
lua package_ui.lua  <pattern> [target]
    <pattern>:
        -load: 仅仅是进行打包生成epp epd文件
        -pack: 生成贴图以及打包成epp epd文件
    [target]:
        common: 默认为此选项, 仅仅是打包common目录下的png 和json
        360: qihoo target
        efun: efun target
]]

local texture_pack = require "texture_pack"
local json = require "ljson"

local join = texture_pack.join
local del_dir = texture_pack.del_dir
local del_file = texture_pack.del_file
local mkdir = texture_pack.mkdir
local tmp_dir = texture_pack.tmp_dir
local run_cmd = texture_pack.run_cmd
local check_have_file = texture_pack.check_have_file


local pattern, target = ...
local target = target or "common"

local epbin = "..\\tools\\lua\\epbin.lua"
local lua = "..\\tools\\lua\\lua52.exe"
local python = "..\\tools\\python27\\python.exe"

local _path_target = {
    ["common"] = "..\\sg_ui\\common",
    ["360"] = "..\\sg_ui\\360",
    ["efun"] = "..\\sg_ui\\efun"
}


--  创建打包环境 type: json/png/all  默认为all
local function _get_target_env(target, type)
    type = type or "all"
    local common_path = _path_target.common
    local target_path = _path_target[target]
    assert(target_path)

    local c_png = common_path.."\\png"
    local c_json = common_path.."\\json"
    local t_png = target_path.."\\png"
    local t_json = target_path.."\\json"

    local ret = {
        png_path = c_png,
        json_path = c_json,
        target_path = "ui_output\\"..target
    }

    -- 检查target文件中是否有png
    if common_path~=target_path and (type=="all" or type=="png") and check_have_file(t_png) then
        tmp_dir('tmp_png')
        run_cmd( ('COPY /Y %s\\*.png "tmp_png\\" '):format(c_png))
        run_cmd( ('COPY /Y %s\\*.png "tmp_png\\" '):format(t_png))
        ret.png_path = 'tmp_png'
    end

    -- 检查target文件是否有json
    if common_path~=target_path and (type=="all" or type=="json") and check_have_file(t_json) then
        tmp_dir('tmp_json')
        run_cmd( ('COPY /Y %s\\*.json "tmp_json\\" '):format(c_json))
        run_cmd( ('COPY /Y %s\\*.json "tmp_json\\" '):format(t_json))
        ret.json_path = 'tmp_json'
    end

    return ret
end

-- 清理打包环境
local function _clean_target_env(target)
    del_dir('tmp_png')
    del_dir('tmp_json')
end

local function _json2table(file)
    local handle = io.open(file)
    local str = handle:read("*a")
    handle:close()
    local success, table =  json.decode(str)
    assert(success)
    return table
end

---------------------- pack ----------------------
local function _gen_pack(target_env)
    local target_path = target_env.target_path
    local png_path = target_env.png_path
    mkdir(target_path)
    return texture_pack.pack(png_path, target_path.."\\ui2")
end

-- 获得五边形能力图的6450-gaoguang-1.png 所在贴图位置
local function _get_pentagon_src(json_file)
    local data = _json2table(json_file).frames
    for i=1,#data do
        if data[i].filename == "6450-gaoguang-1.png" then
            local frame = data[i].frame
            return {
                frame.x+2, 
                frame.y+frame.h-2, 
                frame.x+frame.w-2,
                frame.y+frame.h-2,
                frame.x+frame.w-2,
                frame.y+2,
                frame.x+2,
                frame.y+2
            }
        end
    end
    return false
end

local function _get_pentagon(target_env)
    for i=1,2 do
        local path = target_env.target_path.."\\ui2"..(i)..".json"
        local src = _get_pentagon_src(path)
        if src then
            local src_pos = (('"%d,%d, %d,%d, %d,%d, %d,%d"'):format(
                src[1], src[2],
                src[3], src[4],
                src[5], src[6],
                src[7], src[8]))
            return src_pos, tostring(i-1)
        end
    end
    assert(false)
end

-- 生成角色和英雄的五边形信息
local function _gen_pentagon(target_env)
    local src_pos, tex = _get_pentagon(target_env)
    
    -- gen char
    run_cmd(join(
        python, 
        "../pentagon_import/gen_pentagon.py",
        src_pos, tex, 
        "../pentagon_import/character.csv ../pentagon_import/pt_char.lua 19",
        "10527"))
    
    -- gen hero
    run_cmd(join(
        python,
        "../pentagon_import/gen_pentagon.py",
        src_pos, tex,
        "../pentagon_import/hero.csv ../pentagon_import/pt_hero.lua 7",
        "11527"))
end

-- 合并五边形信息和
local function _gen_merge(target_env)
    local ui2_ori_lua = target_env.target_path.."\\ui2.ori.lua"
    run_cmd("type ..\\pentagon_import\\pt_char.lua >> "..ui2_ori_lua)
    run_cmd("type ..\\pentagon_import\\pt_hero.lua >> "..ui2_ori_lua)
end

local function _gen_load(target_env)
    local target_path = target_env.target_path
    local json_path = target_env.json_path
    mkdir(target_path)
    
    local cmd = join(
        "cocpackage_load.exe",
        json_path, 
        target_path.."\\ui2",
        target_path.."\\ui2.ori.lua"
    )
    run_cmd(cmd)
end

--- 生成EP 文件
local function _gen_ep(target_env)
    local ui2_conver = ".\\ui_output\\ui2.conver.lua"
    local ui2_manual = ".\\ui_output\\ui2.manual.lua"
    local ui2_ori_lua = target_env.target_path.."\\ui2.ori.lua"
    local ui2_lua = target_env.target_path.."\\ui2.lua"

    local ui2 = target_env.target_path.."\\ui2"
    local t_ios = target_env.target_path.."\\ui2.ios"
    local t_android = target_env.target_path.."\\ui2.android"
    
    del_file(ui2_lua)
    
    run_cmd(("type %s >> %s"):format(ui2_conver, ui2_lua))
    run_cmd(("type %s >> %s"):format(ui2_ori_lua, ui2_lua))
    run_cmd(("type %s >> %s"):format(ui2_manual, ui2_lua))
    
    print("---------gen ios ep file-------------")
    run_cmd(("%s %s -pd -png8 %s %s.ios"):format(lua, epbin, ui2_lua, ui2))
    print("---------gen android ep file---------")
    -- fixed me!!!
    run_cmd(("COPY /Y .\\ui_output\\*.pkm %s"):format(target_env.target_path))
    run_cmd(("%s %s -pd -pkmc %s %s.android"):format(lua, epbin, ui2_lua, ui2))
end





----------------------run --------------------------
if not _path_target[target] then
    print(README)
    return
end

if pattern == "-load" then
    local env = _get_target_env(target, "json")
    _gen_pentagon(env)
    _gen_load(env)
    _gen_merge(env)
    _gen_ep(env)
    _clean_target_env(env)
    return
elseif pattern == "-pack" then
    local env = _get_target_env(target)
    _gen_pack(env)
    _gen_pentagon(env)
    _gen_load(env)
    _gen_merge(env)
    _gen_ep(env)
    _clean_target_env(env)
    return
end

print(README)